src/clib/ctfunc.cpp

```diff
@@ -28,78 +26,98 @@ extern "C" {
     int func_new(int type, size_t n, size_t lenp, const double* params)
     {
         try {
-            func_t* r=0;
+            shared_ptr<Func1> r;
             size_t m = lenp;
             if (type == SinFuncType) {
-                r = new Sin1(params[0]);
+                r = newFunc1("sin", params[0]);
             } else if (type == CosFuncType) {
-                r = new Cos1(params[0]);
+                r = newFunc1("cos", params[0]);
             } else if (type == ExpFuncType) {
-                r = new Exp1(params[0]);
+                r = newFunc1("exp", params[0]);
             } else if (type == PowFuncType) {
-                if (lenp < 1) {
-                    throw CanteraError("func_new",
-                                       "exponent for pow must be supplied");
-                }
-                r = new Pow1(params[0]);
+                r = newFunc1("pow", params[0]);
             } else if (type == ConstFuncType) {
-                r = new Const1(params[0]);
+                r = newFunc1("constant", params[0]);
             } else if (type == FourierFuncType) {
-                if (lenp < 2*n + 2) {
-                    throw CanteraError("func_new",
-                                       "not enough Fourier coefficients");
-                }
-                r = new Fourier1(n, params[n+1], params[0], params + 1,
-                                 params + n + 2);
+                vector<double> par(lenp);
+                std::copy(params, params + lenp, par.data());
```

Comment:
Any time you're initializing a `vector` from a `double*`:
```suggestion
                vector<double> par(params, params + lenp);
```